src/docmanager/index.ts

```diff
@@ -0,0 +1,514 @@
+
+import {
+  Widget
+} from 'phosphor-widget';
+
+
+/**
+ * The interface for a document model.
+ */
+export interface IDocumentModel {
+  /**
+   * A signal emitted when the document content changes.
+   */
+  contentChanged: ISignal<IDocumentModel, any>;
+
+  /**
+   * Serialize the model.  It should return a JSON object or a string.
+   */
+  serialize(): any;
+
+  /**
+   * Deserialize the model from a string or a JSON object.
+   *
+   * #### Notes
+   * Should emit a [contentChanged] signal.
+   */
+  deserialize(value: any): void;
+
+  /**
+   * The default kernel name for the document.
+   *
+   * #### Notes
+   * This is a read-only property.
+   */
+  defaultKernelName: string;
+
+  /**
+   * The default kernel language for the document.
+   *
+   * #### Notes
+   * This is a read-only property.
+   */
+  defaultKernelLanguage: string;
+}
+
+
+/**
+ * A session info object for running sessions.
+ */
+export interface ISessionInfo {
+  /**
+   * The list of file paths associated with the running sessions.
+   */
+  paths: string[];
+
+  /**
+   * The list of kernel specs associated with the running sessions.
+   */
+  specs: IKernelSpecId[];
+}
+
+
+/**
+ * The document context object.
+ */
+export interface IDocumentContext {
+  /**
+   * A signal emitted when the kernel changes.
+   */
+  kernelChanged: ISignal<IDocumentContext, IKernel>;
+
+  /**
+   * A signal emitted when the path changes.
+   */
+  pathChanged: ISignal<IDocumentContext, string>;
+
+  /**
+   * A signal emitted when the model is saved or reverted.
+   */
+  dirtyCleared: ISignal<IDocumentContext, void>;
+
+  /**
+   * The current kernel associated with the document.
+   *
+   * #### Notes
+   * This is a read-only property.
+   */
+  kernel: IKernel;
+
+  /**
+   * The current path associated with the document.
+   *
+   * #### Notes
+   * This is a read-only property.
+   */
+  path: string;
+
+  /**
+   * Change the current kernel associated with the document.
+   */
+  changeKernel(options: IKernelId): Promise<IKernel>;
+
+  /**
+   * Save the document contents to disk.
+   */
+  save(): Promise<void>;
+
+  /**
+   * Revert the document contents to disk contents.
+   */
+  revert(): Promise<void>;
+
+  /**
+   * Get the list of running sessions.
+   */
+  listRunningSessions(): Promise<ISessionInfo>;
+
+  /**
+   * Add a sibling widget to the document manager.
+   *
+   * @param widget - The widget to add to the document manager.
+   *
+   * @returns A disposable used to remove the sibling if desired.
+   */
+  addSibling(widget: Widget): IDisposable;
+}
+
+
+/**
+ * The interface for a widget factory.
+ */
+export
+interface IWidgetFactory<T extends Widget> {
+  /**
+   * Create a new widget given a document model and a context.
+   */
+  createNew(model: IDocumentModel, context: IDocumentContext): T;
+
+  /**
+   * Get the default kernel name given a path and a list of specs.
+   */
+  getDefaultKernel(path: string, specs: IKernelSpecId[]): string;
```

Comment:
We have this and the getPreferredKernels method so that when we are doing an openWith or create, we can select a widget and then see possible kernels for the filename. But really that decision is made at the model level already. To make things consistent, I suppose we should construct the model, hand it the path, then get back the preferred kernel name and language, then sort the kernel specs for that preferred kernel name and language. But at that point, I think we should do that logic outside of IWidgetFactory.

So I don't think IWidgetFactory needs this and the getPreferredKernels functions.